core/signer/signer.go

```diff
@@ -63,10 +63,26 @@ func New(realm string,
 	}
 }
 
-// StartSigning advance epochs once per interval.
-func (s *Signer) StartSigning(ctx context.Context, interval time.Duration) {
-	for range time.NewTicker(interval).C {
-		if err := s.CreateEpoch(ctx); err != nil {
+// StartSigning advance epochs once per minInterval, if there were mutations, and at least once per
+// maxElapsed minIntervals.
+func (s *Signer) StartSigning(ctx context.Context, minInterval time.Duration, maxInterval time.Duration) {
+	tc := time.Tick(minInterval)
+	processEpochs(ctx, tc, maxInterval, s.CreateEpoch)
+}
+
+func processEpochs(ctx context.Context, tc <-chan time.Time, maxElapsed time.Duration, sign func(ctx context.Context, enforce bool) error) {
+	last := time.Now()
+	for now := range tc {
+		var forceNewEpoch bool
+		if time.Since(last) >= maxElapsed {
+			forceNewEpoch = true
+			last = now
+		} else {
+			forceNewEpoch = false
+		}
+		// only create a new epoch if elapsed intervals >= max. elapsed intervals
+		// OR there were mutations in between max. elapsed epochs (expected to be handled by passed callback):
+		if err := sign(ctx, forceNewEpoch); err != nil {
```

Comment:
Indeed, this is a call to `CreateEpoch`. `sign` is a "callback" passed to `processEpochs` (and `s.CreateEpoch` is passed as an argument).
Again, this only exists to make it testable without having to run a whole environment (like for instance in this [integration](https://github.com/google/keytransparency/blob/master/integration/testutil.go#L154-L213) package).